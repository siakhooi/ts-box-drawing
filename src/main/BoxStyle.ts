export class BoxStyle {
  HORIZONTAL: string;
  VERTICAL: string;
  TOP_LEFT: string;
  TOP_RIGHT: string;
  BOTTOM_LEFT: string;
  BOTTOM_RIGHT: string;
  CROSS: string;
  TOP_MIDDLE: string;
  BOTTOM_MIDDLE: string;
  MIDDLE_LEFT: string;
  MIDDLE_RIGHT: string;

  constructor(
    horizontal: string,
    vertical: string,
    top_left: string,
    top_right: string,
    bottom_left: string,
    bottom_right: string,
    cross: string,
    top_middle: string,
    bottom_middle: string,
    middle_left: string,
    middle_right: string
  ) {
    this.HORIZONTAL = horizontal;
    this.VERTICAL = vertical;
    this.TOP_LEFT = top_left;
    this.TOP_RIGHT = top_right;
    this.BOTTOM_LEFT = bottom_left;
    this.BOTTOM_RIGHT = bottom_right;
    this.CROSS = cross;
    this.TOP_MIDDLE = top_middle;
    this.BOTTOM_MIDDLE = bottom_middle;
    this.MIDDLE_LEFT = middle_left;
    this.MIDDLE_RIGHT = middle_right;
  }
}
