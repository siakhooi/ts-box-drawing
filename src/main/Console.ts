import util from 'util';

class ConsoleClass {
  println(format: string, ...texts: string[]): void {
    console.log(util.format(format, ...texts));
  }
}

export const Console = new ConsoleClass();
