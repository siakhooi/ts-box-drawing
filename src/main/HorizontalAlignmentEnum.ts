export enum HorizontalAlignmentEnum {
  LEFT,
  CENTER,
  RIGHT,
}
