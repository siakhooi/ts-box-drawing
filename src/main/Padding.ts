export class Padding {
  private padByColumn: number[] = [];
  private defaultPad = 0;

  setDefault(number_of_spaces: number) {
    this.defaultPad = number_of_spaces;
  }
  setPadding(column: number, number_of_spaces: number) {
    this.padByColumn[column] = number_of_spaces;
  }
  getPad(column: number): number {
    return this.padByColumn[column] === undefined
      ? this.defaultPad
      : this.padByColumn[column];
  }
  padAll(numbers_of_spaces: number[]) {
    numbers_of_spaces.forEach((value, index) => {
      this.setPadding(index, value);
    });
  }
}
