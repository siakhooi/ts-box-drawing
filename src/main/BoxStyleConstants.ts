import {BoxStyle} from './BoxStyle';

const THIN = new BoxStyle(
  '\u2500',
  '\u2502',
  '\u250c',
  '\u2510',
  '\u2514',
  '\u2518'
);
const THICK = new BoxStyle(
  '\u2501',
  '\u2503',
  '\u250F',
  '\u2513',
  '\u2517',
  '\u251B'
);
const DOUBLE = new BoxStyle(
  '\u2550',
  '\u2551',
  '\u2554',
  '\u2557',
  '\u255A',
  '\u255D'
);
const THINCURVE = new BoxStyle(
  '\u2500',
  '\u2502',
  '\u256D',
  '\u256E',
  '\u2570',
  '\u256F'
);

export {THIN, THICK, DOUBLE, THINCURVE};
