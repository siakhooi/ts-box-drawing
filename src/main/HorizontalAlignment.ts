import {HorizontalAlignmentEnum} from './HorizontalAlignmentEnum';

export class HorizontalAlignment {
  private alignmentByColumn: HorizontalAlignmentEnum[] = [];
  private defaultAlignment = HorizontalAlignmentEnum.LEFT;

  getLine(column: number, cellData: string, columnWidth: number) {
    const left = (columnWidth - cellData.length) / 2;

    switch (this.getAlignment(column)) {
      case HorizontalAlignmentEnum.LEFT:
        return cellData.padEnd(columnWidth, ' ');
      case HorizontalAlignmentEnum.RIGHT:
        return cellData.padStart(columnWidth, ' ');
      case HorizontalAlignmentEnum.CENTER:
        return (' '.repeat(left) + cellData).padEnd(columnWidth, ' ');
    }
  }

  setAlign(column: number, alignment: HorizontalAlignmentEnum) {
    this.alignmentByColumn[column] = alignment;
  }
  private getAlignment(column: number): HorizontalAlignmentEnum {
    return this.alignmentByColumn[column] === undefined
      ? this.defaultAlignment
      : this.alignmentByColumn[column];
  }
}
