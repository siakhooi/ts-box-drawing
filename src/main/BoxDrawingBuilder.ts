import {BoxStyle} from './BoxStyle';
import {BoxStyles} from './BoxStyles';
import {Console} from './Console';
import {Padding} from './Padding';

type BoxDataType = null | string | string[][];

function getColumnWidths(
  rowCount: number,
  columnCount: number,
  data: string[][]
): number[] {
  const columnWidths: number[] = [];
  for (let i = 0; i < columnCount; i++) {
    columnWidths[i] = 0;
    for (let j = 0; j < rowCount; j++) {
      if (data[j][i] !== undefined) {
        columnWidths[i] = Math.max(columnWidths[i], data[j][i].length);
      }
    }
  }
  return columnWidths;
}

function convertDataToArray(data: BoxDataType): string[][] {
  if (data === '' || data === null) return [];
  return typeof data === 'string' ? [[data]] : data;
}

export class BoxDrawingBuilder {
  private data: BoxDataType = [];
  private style = BoxStyles.THIN;

  private paddingLeft = new Padding();
  private paddingRight = new Padding();

  setData(data: BoxDataType): BoxDrawingBuilder {
    this.data = data;
    return this;
  }
  setStyle(style: BoxStyle): BoxDrawingBuilder {
    this.style = style;
    return this;
  }
  setDefaultPadLeft(number_of_spaces: number): BoxDrawingBuilder {
    this.paddingLeft.setDefault(number_of_spaces);
    return this;
  }
  setDefaultPadRight(number_of_spaces: number): BoxDrawingBuilder {
    this.paddingRight.setDefault(number_of_spaces);
    return this;
  }
  padLeftAll(numbers_of_spaces: number[]): BoxDrawingBuilder {
    this.paddingLeft.padAll(numbers_of_spaces);
    return this;
  }
  padLeft(column: number, number_of_spaces: number): BoxDrawingBuilder {
    this.paddingLeft.setPadding(column, number_of_spaces);
    return this;
  }
  padRight(column: number, number_of_spaces: number): BoxDrawingBuilder {
    this.paddingRight.setPadding(column, number_of_spaces);
    return this;
  }
  pad(
    column: number,
    number_of_spaces_on_left: number,
    number_of_spaces_on_right: number
  ): BoxDrawingBuilder {
    this.padLeft(column, number_of_spaces_on_left);
    this.padRight(column, number_of_spaces_on_right);
    return this;
  }
  private getHorizontalLines(
    columnCount: number,
    columnWidths: number[]
  ): string[] {
    let lineTop = '',
      lineMiddle = '',
      lineBottom = '';
    const boxStyle = this.style;

    lineTop = boxStyle.TOP_LEFT;
    lineMiddle = boxStyle.MIDDLE_LEFT;
    lineBottom = boxStyle.BOTTOM_LEFT;

    for (let i = 0; i < columnCount; i++) {
      const padLeft = this.paddingLeft.getPad(i);
      const padRight = this.paddingRight.getPad(i);
      const columnWidth = columnWidths[i] + padLeft + padRight;
      lineTop += boxStyle.HORIZONTAL_OUTER.repeat(columnWidth);
      lineMiddle += boxStyle.HORIZONTAL_INNER.repeat(columnWidth);
      lineBottom += boxStyle.HORIZONTAL_OUTER.repeat(columnWidth);

      if (i < columnCount - 1) {
        lineTop += boxStyle.TOP_MIDDLE;
        lineMiddle += boxStyle.CROSS;
        lineBottom += boxStyle.BOTTOM_MIDDLE;
      }
    }
    lineTop += boxStyle.TOP_RIGHT;
    lineMiddle += boxStyle.MIDDLE_RIGHT;
    lineBottom += boxStyle.BOTTOM_RIGHT;

    return [lineTop, lineMiddle, lineBottom];
  }
  private getDataLine(
    columnCount: number,
    columnWidths: number[],
    rowData: string[]
  ): string {
    const boxStyle = this.style;
    let s = boxStyle.VERTICAL_OUTER;
    for (let i = 0; i < columnCount; i++) {
      const padLeft = this.paddingLeft.getPad(i);
      const padRight = this.paddingRight.getPad(i);
      const cellData = rowData[i] === undefined ? '' : rowData[i];
      s += ' '.repeat(padLeft);
      s += cellData.padEnd(columnWidths[i], ' ');
      s += ' '.repeat(padRight);
      if (i < columnCount - 1) s += boxStyle.VERTICAL_INNER;
    }
    s += boxStyle.VERTICAL_OUTER;

    return s;
  }
  drawBox() {
    const data: string[][] = convertDataToArray(this.data);
    const rowCount = data.length;
    const columnCount = data.reduce((r, n) => Math.max(r, n.length), 0);

    const columnWidths: number[] = getColumnWidths(rowCount, columnCount, data);

    let lineTop = '',
      lineMiddle = '',
      lineBottom = '';
    [lineTop, lineMiddle, lineBottom] = this.getHorizontalLines(
      columnCount,
      columnWidths
    );

    Console.println(lineTop);

    for (let j = 0; j < rowCount; j++) {
      Console.println(this.getDataLine(columnCount, columnWidths, data[j]));
      if (j < rowCount - 1) Console.println(lineMiddle);
    }

    Console.println(lineBottom);
  }
}
