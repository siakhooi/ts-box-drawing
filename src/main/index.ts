import {BoxStyle} from './BoxStyle';
import {Console} from './Console';
import {THIN, THICK, DOUBLE, THINCURVE} from './BoxStyleConstants';

function drawBox(boxStyle: BoxStyle, text: string) {
  const w = text.length;
  Console.printf(
    '%s%s%s',
    boxStyle.TOP_LEFT,
    boxStyle.HORIZONTAL.repeat(w),
    boxStyle.TOP_RIGHT
  );
  Console.printf('%s%s%s', boxStyle.VERTICAL, text, boxStyle.VERTICAL);
  Console.printf(
    '%s%s%s',
    boxStyle.BOTTOM_LEFT,
    boxStyle.HORIZONTAL.repeat(w),
    boxStyle.BOTTOM_RIGHT
  );
}
export function drawBoxThin(text: string) {
  drawBox(THIN, text);
}
export function drawBoxThick(text: string) {
  drawBox(THICK, text);
}
export function drawBoxDouble(text: string) {
  drawBox(DOUBLE, text);
}
export function drawBoxThinCurve(text: string) {
  drawBox(THINCURVE, text);
}
