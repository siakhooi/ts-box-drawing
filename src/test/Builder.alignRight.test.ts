import {Console} from '../main/Console';
import {BoxDrawingBuilder} from '../main/index';
import {
  ALIGN_TEST_DATA,
  ALL_STYLES,
  TestDataUnit,
  TestStyleUnit,
} from './TestData';

describe.each(ALL_STYLES)('AlignRight([2,3])', (testStyle: TestStyleUnit) => {
  test.each(ALIGN_TEST_DATA)(
    `${testStyle.title} [$title]`,
    (testData: TestDataUnit) => {
      const allData: string[] = [];
      Console.println = jest
        .fn()
        .mockImplementation(text => allData.push(text));
      new BoxDrawingBuilder()
        .setData(testData.data)
        .setStyle(testStyle.style)
        .alignRight([2, 3])
        .drawBox();
      expect(allData).toMatchSnapshot();
    }
  );
});
