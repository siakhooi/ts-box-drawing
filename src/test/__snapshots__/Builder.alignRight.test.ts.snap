// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlignRight([2,3]) DOUBLE [Simple 2x2] 1`] = `
[
  "╔════════╦════════╦═════════╦════════╗",
  "║Column 1║Column 2║Column 3A║Column 4║",
  "╠════════╬════════╬═════════╬════════╣",
  "║Left    ║Ctr     ║      Ctr║   Right║",
  "╠════════╬════════╬═════════╬════════╣",
  "║Left 1  ║Ctr2    ║     Ctr3║ Right 4║",
  "╚════════╩════════╩═════════╩════════╝",
]
`;

exports[`AlignRight([2,3]) HORIZONTAL_DOUBLE_VERTICAL_THIN [Simple 2x2] 1`] = `
[
  "╒════════╤════════╤═════════╤════════╕",
  "│Column 1│Column 2│Column 3A│Column 4│",
  "╞════════╪════════╪═════════╪════════╡",
  "│Left    │Ctr     │      Ctr│   Right│",
  "╞════════╪════════╪═════════╪════════╡",
  "│Left 1  │Ctr2    │     Ctr3│ Right 4│",
  "╘════════╧════════╧═════════╧════════╛",
]
`;

exports[`AlignRight([2,3]) HORIZONTAL_THICK_VERTICAL_THIN [Simple 2x2] 1`] = `
[
  "┍━━━━━━━━┯━━━━━━━━┯━━━━━━━━━┯━━━━━━━━┑",
  "│Column 1│Column 2│Column 3A│Column 4│",
  "┝━━━━━━━━┿━━━━━━━━┿━━━━━━━━━┿━━━━━━━━┥",
  "│Left    │Ctr     │      Ctr│   Right│",
  "┝━━━━━━━━┿━━━━━━━━┿━━━━━━━━━┿━━━━━━━━┥",
  "│Left 1  │Ctr2    │     Ctr3│ Right 4│",
  "┕━━━━━━━━┷━━━━━━━━┷━━━━━━━━━┷━━━━━━━━┙",
]
`;

exports[`AlignRight([2,3]) HORIZONTAL_THIN_VERTICAL_DOUBLE [Simple 2x2] 1`] = `
[
  "╓────────╥────────╥─────────╥────────╖",
  "║Column 1║Column 2║Column 3A║Column 4║",
  "╟────────╫────────╫─────────╫────────╢",
  "║Left    ║Ctr     ║      Ctr║   Right║",
  "╟────────╫────────╫─────────╫────────╢",
  "║Left 1  ║Ctr2    ║     Ctr3║ Right 4║",
  "╙────────╨────────╨─────────╨────────╜",
]
`;

exports[`AlignRight([2,3]) HORIZONTAL_THIN_VERTICAL_THICK [Simple 2x2] 1`] = `
[
  "┎────────┰────────┰─────────┰────────┒",
  "┃Column 1┃Column 2┃Column 3A┃Column 4┃",
  "┠────────╂────────╂─────────╂────────┨",
  "┃Left    ┃Ctr     ┃      Ctr┃   Right┃",
  "┠────────╂────────╂─────────╂────────┨",
  "┃Left 1  ┃Ctr2    ┃     Ctr3┃ Right 4┃",
  "┖────────┸────────┸─────────┸────────┚",
]
`;

exports[`AlignRight([2,3]) INNER_DOUBLE_OUTER_THIN [Simple 2x2] 1`] = `
[
  "┌────────╥────────╥─────────╥────────┐",
  "│Column 1║Column 2║Column 3A║Column 4│",
  "╞════════╬════════╬═════════╬════════╡",
  "│Left    ║Ctr     ║      Ctr║   Right│",
  "╞════════╬════════╬═════════╬════════╡",
  "│Left 1  ║Ctr2    ║     Ctr3║ Right 4│",
  "└────────╨────────╨─────────╨────────┘",
]
`;

exports[`AlignRight([2,3]) INNER_DOUBLE_OUTER_THIN_CURVE [Simple 2x2] 1`] = `
[
  "╭────────╥────────╥─────────╥────────╮",
  "│Column 1║Column 2║Column 3A║Column 4│",
  "╞════════╬════════╬═════════╬════════╡",
  "│Left    ║Ctr     ║      Ctr║   Right│",
  "╞════════╬════════╬═════════╬════════╡",
  "│Left 1  ║Ctr2    ║     Ctr3║ Right 4│",
  "╰────────╨────────╨─────────╨────────╯",
]
`;

exports[`AlignRight([2,3]) INNER_THICK_OUTER_THIN [Simple 2x2] 1`] = `
[
  "┌────────┰────────┰─────────┰────────┐",
  "│Column 1┃Column 2┃Column 3A┃Column 4│",
  "┝━━━━━━━━╋━━━━━━━━╋━━━━━━━━━╋━━━━━━━━┥",
  "│Left    ┃Ctr     ┃      Ctr┃   Right│",
  "┝━━━━━━━━╋━━━━━━━━╋━━━━━━━━━╋━━━━━━━━┥",
  "│Left 1  ┃Ctr2    ┃     Ctr3┃ Right 4│",
  "└────────┸────────┸─────────┸────────┘",
]
`;

exports[`AlignRight([2,3]) INNER_THICK_OUTER_THIN_CURVE [Simple 2x2] 1`] = `
[
  "╭────────┰────────┰─────────┰────────╮",
  "│Column 1┃Column 2┃Column 3A┃Column 4│",
  "┝━━━━━━━━╋━━━━━━━━╋━━━━━━━━━╋━━━━━━━━┥",
  "│Left    ┃Ctr     ┃      Ctr┃   Right│",
  "┝━━━━━━━━╋━━━━━━━━╋━━━━━━━━━╋━━━━━━━━┥",
  "│Left 1  ┃Ctr2    ┃     Ctr3┃ Right 4│",
  "╰────────┸────────┸─────────┸────────╯",
]
`;

exports[`AlignRight([2,3]) INNER_THIN_OUTER_DOUBLE [Simple 2x2] 1`] = `
[
  "╔════════╤════════╤═════════╤════════╗",
  "║Column 1│Column 2│Column 3A│Column 4║",
  "╟────────┼────────┼─────────┼────────╢",
  "║Left    │Ctr     │      Ctr│   Right║",
  "╟────────┼────────┼─────────┼────────╢",
  "║Left 1  │Ctr2    │     Ctr3│ Right 4║",
  "╚════════╧════════╧═════════╧════════╝",
]
`;

exports[`AlignRight([2,3]) INNER_THIN_OUTER_THICK [Simple 2x2] 1`] = `
[
  "┏━━━━━━━━┯━━━━━━━━┯━━━━━━━━━┯━━━━━━━━┓",
  "┃Column 1│Column 2│Column 3A│Column 4┃",
  "┠────────┼────────┼─────────┼────────┨",
  "┃Left    │Ctr     │      Ctr│   Right┃",
  "┠────────┼────────┼─────────┼────────┨",
  "┃Left 1  │Ctr2    │     Ctr3│ Right 4┃",
  "┗━━━━━━━━┷━━━━━━━━┷━━━━━━━━━┷━━━━━━━━┛",
]
`;

exports[`AlignRight([2,3]) THICK [Simple 2x2] 1`] = `
[
  "┏━━━━━━━━┳━━━━━━━━┳━━━━━━━━━┳━━━━━━━━┓",
  "┃Column 1┃Column 2┃Column 3A┃Column 4┃",
  "┣━━━━━━━━╋━━━━━━━━╋━━━━━━━━━╋━━━━━━━━┫",
  "┃Left    ┃Ctr     ┃      Ctr┃   Right┃",
  "┣━━━━━━━━╋━━━━━━━━╋━━━━━━━━━╋━━━━━━━━┫",
  "┃Left 1  ┃Ctr2    ┃     Ctr3┃ Right 4┃",
  "┗━━━━━━━━┻━━━━━━━━┻━━━━━━━━━┻━━━━━━━━┛",
]
`;

exports[`AlignRight([2,3]) THIN [Simple 2x2] 1`] = `
[
  "┌────────┬────────┬─────────┬────────┐",
  "│Column 1│Column 2│Column 3A│Column 4│",
  "├────────┼────────┼─────────┼────────┤",
  "│Left    │Ctr     │      Ctr│   Right│",
  "├────────┼────────┼─────────┼────────┤",
  "│Left 1  │Ctr2    │     Ctr3│ Right 4│",
  "└────────┴────────┴─────────┴────────┘",
]
`;

exports[`AlignRight([2,3]) THINCURVE [Simple 2x2] 1`] = `
[
  "╭────────┬────────┬─────────┬────────╮",
  "│Column 1│Column 2│Column 3A│Column 4│",
  "├────────┼────────┼─────────┼────────┤",
  "│Left    │Ctr     │      Ctr│   Right│",
  "├────────┼────────┼─────────┼────────┤",
  "│Left 1  │Ctr2    │     Ctr3│ Right 4│",
  "╰────────┴────────┴─────────┴────────╯",
]
`;

exports[`AlignRight([2,3]) VERTICALBAR_PLUS_MINUS [Simple 2x2] 1`] = `
[
  "+--------+--------+---------+--------+",
  "|Column 1|Column 2|Column 3A|Column 4|",
  "+--------+--------+---------+--------+",
  "|Left    |Ctr     |      Ctr|   Right|",
  "+--------+--------+---------+--------+",
  "|Left 1  |Ctr2    |     Ctr3| Right 4|",
  "+--------+--------+---------+--------+",
]
`;
